### ISSUE TITLE
get_css() takes 3 positional arguments but 4 were given

### ISSUE THREAD

**Issue by dexedge** [donhauser/wagtail-pdf#7]
Hi,

I'm trying to add `wagtail-pdf` to a site that is just about ready to launch. I tried to add pdf generation to the main app for the site, but I may have misunderstood how to do this. What I did was simply to have my existing Page model inherit from `PdfViewPageMixin`, and to add `ROUTE_CONFIG`to the model, as below:

```
class DocumentPage(PdfViewPageMixin, Page):
    template  = "documents/document.html"
    parent_page_types = ["top.IndexPage", "DocumentList"]
    
    date = models.DateField("Date", null=True)
    date_precision = models.CharField(

[snip...]

    ROUTE_CONFIG = [
        ("html", r'^$'),
        ("pdf", r'^pdf/$'),
    ]
```
The "PDF Preview" option shows up correctly in the Wagtail admin interface, but when I select it, I get the error `get_css() takes 3 positional arguments but 4 were given`. I get the same error when I add /pdf to the end of the URL for the published page. The HTML page still renders correctly.

Can someone help me figure out what I'm doing wrong?

Thanks!

**Comment by donhauser** [comment]
Hi,
cool that you are using this library!
I did not have that error before I can't directly see something in your code that is wrong, so I would guess that this is a bug.

While I just wrote this, I did a quick search on the dependencies of `get_css()` and I found that `django-weasyprint` has [altered this method](https://github.com/fdemmer/django-weasyprint/commit/f944d4501e7e864a48ab9389920ce089d6823b01#diff-8f42ee01bb6e6cbe9799c6cecd12a876acd727cba649717e1cfe97b58c7ce795)
In particular this library still uses `get_css(self, base_url, url_fetcher)`, while the latest version of that library requires `get_css(self, base_url, url_fetcher, font_config)` which would perfectly match to your reported failure.

To resolve it simply downgrade `django-weasyprint` to a version that was created before october 21.

Hope that helps & Cheers!

**Comment by dexedge** [comment]
Thanks! I'll try that and report back.

**Comment by dexedge** [comment]
First attempt:

I uninstalled `django-weasyprint 2.1.0` (the current version) and `weasyprint` 54.0, then installled `django-weasyprint` 2.0 (which installed `weasyprint` 53.0). Now I get a different error:

```function/symbol 'pango_context_set_round_glyph_positions' not found in library '/usr/local/lib/libpango-1.0.dylib': dlsym(0x7f9b9d413720, pango_context_set_round_glyph_positions): symbol not found```

I'll try going back another version to see if that helps.

**Comment by dexedge** [comment]
I installed `django-weasyprint` 1.1.0.post1, the last version before 2.0. Now it doesn't crash, but all I get with `PDF Preview` in the `wagtail` admin (and with `/pdf` at the end of the page URL) is a single page with a rectangle of the background color of my site. 

But tomorrow I will try a template dedicated to the pdf to see what happens.

I should have mentioned that I'm using `wagtail` 2.15.1 and `Django` 3.2.7.

**Comment by donhauser** [comment]
Hm okay,
from my experience the `pango` errors are among the most annoying ones, `pango` itself is a dependency of `weasyprint` that is used under the hood to render the pdfs.
I've had the case that `pango` (which is a C library) was highly outdated on the webserver and nothing did work.
In particular updating `weasyprint` via pip can't solve those issues as `pango` needs to be upgraded by the systems package manager (e.g. `apt-get`).
So before you go down all that struggle I would recommend  that you check whether `pango` is using a recent version, this really saves a lot of frustration ;)

Your last report does not sound too bad, you need to make sure to use a dedicated pdf template if you are using a mixed html/pdf side. I would guess that `weasyprint` tries to render the full html page and that will not work out.

Maybe a good first step would be to fork the `demo` folder of this project and to see whether it works out as it's supposed to. Then you at least know, whether you have some package issues or just some mistake in your code.

**Comment by dexedge** [comment]
Thanks for your help. I tried to incorporate elements into my existing model from `HtmlAndPdfPage` in `home.models` and `pdf_document_base.html` from the demo, but so far no luck. I may not correctly understand how this is supposed to work.

**Comment by donhauser** [comment]
Sure, no problem!

In general creating a "plain" PDF page should be relatively easy, you just need to ensure that `weasyprint` will mess up, if you simply pass the HTML of an entire website to it.

Especially if you using the shared HTML&PDF mode you need to take care of this.

The demo uses `document_base.html` as base HTML-page layout and `pdf_document_base.html` as PDF-page layout and then inserts the actual page content into both templates, such that your browser and `weasyprint` both get their own distinctive HTML-code.
In particular, `pdf_document_base.html` does not extend the django base template here, such that all the JS stuff and so on is not included.
```{% extends 'base.html' %}```

Sorry that you are going through so much trouble here, usually this is really rather plug and play ;)

I probably need to add more details on this to the project

**Comment by dexedge** [comment]
Thanks and no problem. I'd like to be able to use your package, so I'd like to do whatever I can to help figure out why it's not working.

I thought that I had already understood everything you wrote in your last response, but I hadn't been able to get it to work. I uninstalled everything, took out all of the relevant code, and today I started over from scratch. I'll give my detailed steps, in case I'm missing something obvious.

- I installed `wagtail-pdf-view` 0.2.0 and `django-weasyprint` 1.1.0.post1 (which installs `weasyprint` 52.5).
- I added `wagtail_pdf_view` to my `INSTALLED_APPS` (my project was already using `wagtail.contrib.routable_page`, so that was already in `INSTALLED_APPS`)
- In `models.py` for the app (`documents`) I'm trying to add pdfs to, I added `from wagtail_pdf_view.mixins import PdfViewPageMixin`
- The existing model I'm trying to add pdfs to is `DocumentPage`. I modified that model using your `HtmlAndPdfPage` as my guide:

```
class DocumentPage(PdfViewPageMixin, Page):
    template  = "documents/document.html"

[snip...]

# HTML first
    ROUTE_CONFIG = [
        ("html", r'^$'),
        ("pdf", r'^pdf/$'),
    ]

[snip]

    pdf_base_template = "document_pdf.html"

    def get_context(self, request, mode=None, **kwargs):
        context = super().get_context(request, **kwargs)
        
        if mode == 'pdf':
            context["override_base"] = self.pdf_base_template
        
        return context

[snip]
```
I created `document_pdf.html` to be as simple as possible, just to see if Wagtail is even finding it;

```
{% load wagtailcore_tags %}
<!DOCTYPE html>
<html lang="">
  <head>
    <meta charset="utf-8">
    <title>{{page.title}}</title>
    
  </head>
  <body>    
    <h1>Hello world</h1>  
  </body>
</html>
```
In the Wagtail editor, Preview now offers "PDF Preview" as a choice (as expected). "HTML Preview" gives a preview using my normal `base.html` and `document.html` at `http://127.0.0.1:8000/admin/pages/26/edit/preview/?mode=html`, also as expected. 

"PDF Preview" opens `http://127.0.0.1:8000/admin/pages/26/edit/preview/?mode=pdf`, but it seems not to find my `document_pdf.html`, which is currently at the top level of my `templates` folder (I've also tried it in the `documents` subfolder under `templates`). It opens a page in Firefox's pdf reader, but it contains only a rectangle in the content area that is the color of the background from my site (`#f6f7f6`). This is a surprise, because I can't see why it would be taking anything at all from the main CSS for my site. The "Hello world"  text doesn't appear.

No worries if you haven't got the time to diagnose this now. I just wanted you to have the full information. Let me know if there is anything else I can tell you.

**Comment by donhauser** [comment]
Okay now I think we are getting there!
Again its still looks like you are using this correctly but I suspect that there is a small part missing in your `templates/documents/document.html` file:

Since you do still get the gray `#f6f7f6` box and **not** a `TemplateDoesNotExist` error, django does not hook in your PDF-template at all.
From your code, I can see that you've cloned the `get_context()` method from the demo, which sets the `override_base` variable in the context.
Importantly, this context variable needs to be incorporated into your `document.html` (`html_and_pdf_page.html` in the demo)
```html
{% extends override_base|default:'document_base.html' %}
```

By this you can selectively embed the content of `document.html` into either `pdf_document_base.html` (for PDF) and `document_base.html` (for HTML)

*For the sake of completeness:* If your PDF content is supposed to be very different from your HTML page content (*e.g. the PDF renders a menu and while HTML page is a delivery service form*), you should rather hook in entirely separated templates:
You can simply override `PdfViewPageMixin.get_template()` in your Page Model (**important** I did not test this snippet):
```py
class DocumentPage(PdfViewPageMixin, Page):
    # override the default template selection
    def get_template(self, request, extension=None, view_provider=None, **kwargs):
        
        # Check whether a PDF was requested or not
        if isinstance(view_provider, PdfViewPageMixin.PDF_VIEW_PROVIDER):
            return 'path/to/html/template.html'

        return 'path/to/html/template.html'
```

Even more info that you wont probably need, but might be relevant to others: The `get_template()` is called by `WagtailAdapterMixin.get_template_names()`, from which `WagtailWeasyView` inherits. On the other hand `WagtailWeasyView` is specified as `DEFAULT_PDF_VIEW_PROVIDER` and (assuming you did not override this in the settings), `PdfViewPageMixin.PDF_VIEW_PROVIDER` is just `DEFAULT_PDF_VIEW_PROVIDER`.

Hope this suffies for you to get the job done nicely now!
During the next weeks I will add simplification to the two presented methods, this is still to complicated and needs to be more intuitive.
Besides that I will add a fix for the latest weasyprint and django-tex :+1: in the next days

**Comment by dexedge** [comment]
> Importantly, this context variable needs to be incorporated into your document.html (html_and_pdf_page.html in the demo)

>{% extends override_base|default:'document_base.html' %}

>By this you can selectively embed the content of document.html into either pdf_document_base.html (for PDF) and document_base.html (for HTML)

Aha! That's what I overlooked! Now it works!

I guess maybe that ought to be in the README ;-)

I'd be happy to help with the project. My coding skills may not be advanced enough (yet, at least) to help with the code, but I'd be happy to help with the README.

Thanks so much for your help and patience!

**Comment by donhauser** [comment]
Okay nice that it works now!

Contributions of any kind are very welcome, if you want to participate besides coding:
- you can help by adding details to the README, (*which clearly lacks some information here..*)
- by reporting bugs and non-intuitive stuff of all kind
- or by improving the demo

In case of the README, just add your adjustments and make a pull request

Cheers! :)

**Comment by donhauser** [comment]
> The "PDF Preview" option shows up correctly in the Wagtail admin interface, but when I select it, I get the error `get_css() takes 3 positional arguments but 4 were given`. I get the same error when I add /pdf to the end of the URL for the published page. The HTML page still renders correctly.

This is fixed now in version [0.2.1](https://github.com/donhauser/wagtail-pdf/releases/tag/0.2.1)

**Comment by dexedge** [comment]
Just installed 0.2.1, and everything seems to be working smoothly.

**Comment by dexedge** [comment]
Sorry, I spoke too soon. 0.2.1  seems to have broken pdf generation for me. I reverted to 0.2.0, and it works (with wagtail 2.15.3, django-weasyprint 1.1.0.post1, and WeasyPrint 52.5.

**Comment by donhauser** [comment]
> Sorry, I spoke too soon. 0.2.1 seems to have broken pdf generation for me. I reverted to 0.2.0, and it works (with wagtail 2.15.3, django-weasyprint 1.1.0.post1, and WeasyPrint 52.5.

Did you just update this library, or *django-weasyprint* and *weasyprint* as well?
I've just tested 0.2.1 against the latest versions of the related libraries and I think the changes are not compatible to the libraries previous versions.

Just in case that you are using *Debian 10 (buster)*: The latest *weasyprint* requires *Debian 11* (see https://github.com/donhauser/wagtail-pdf/issues/8)

If upgrading does not fix your problem with 0.2.1, it would be nice if you could add some further information (error message, ..) on this :+1:

**Comment by dexedge** [comment]
It turned out to be a pango problem, as you had mentioned might be the case. This afternoon I installed the current stable versions of everything: Weasyprint 54.1, django-weasyprint 2.1.0, and wagtail-pdf-view 0.2.1. This crashed on my local machine with a pango-related error. But I pushed it anyway to the live version of my site on Heroku, and it worked fine there. So I updated pango to 1.50 on my Mac via homebrew (which took about an hour!), and now everything seems to work locally too. Fingers crossed!
